[ffi] add HL_ prefix and move outside of hl.h, allow disable legacy definition

yuxiaomao opened the pull request:
See
- #807 

As suggested added `HL_` prefix and `HL_DISABLE_LEGACY_FFI`

yuxiaomao committed:
FFI: isolate ffi outside of hl.h

yuxiaomao committed:
Do not auto include hl_ffi.h in hl.h if disable legacy

yuxiaomao committed:
Add hl_ffi.h in release, cleanup vcxproj ffi not used

tobil4sk commented:
I wonder if it might make sense to do something like this:

hl.h contains legacy ffi definitions only, instead of them being in hl_ffi.h and hl.h including hl_ffi.h. hl_ffi.h only would contain the new definitions, so it is clean. Also hl_ffi.h can define `HL_DISABLE_LEGACY_FFI` before including hl.h. This way if users migrate to hl_ffi.h they don't have to worry about the bad macro definitions like `_GUID`, but if they include hl.h directly (but without  `HL_DISABLE_LEGACY_FFI`) they still get legacy definitions as before.

We should also add `HL_DISABLE_LEGACY_FFI` to hlc templates since ffi definitions are not needed to compile hlc code and it reduces name clashes.

yuxiaomao committed:
Add -D HL_DISABLE_LEGACY_FFI to hlc template

yuxiaomao committed:
Merge remote-tracking branch 'origin/master' into dev-ffi

tobil4sk commented:
Hm, you're right that's a good point, it's annoying when header order changes behaviour like that.

I think it would still make sense to keep the legacy definitions out of hl_ffi.h. Also, I think you shouldn't get the new definitions unless you include `hl_ffi.h`, otherwise the migration is a bit confusing.

If we reorder a few of the definitions we can make hl_ffi.h independent of hl.h. For example, I think `HL_NO_OPT` isn't really cffi related so it can just go in `hl.h`, and the `vstring` declaration as well. This way `hl_ffi.h` does not have to include `hl.h`. I think this is better as it makes the purpose of each header more concrete.

yuxiaomao committed:
Keep legacy definitions in hl.h, fix HL_STRING

yuxiaomao committed:
Revert "Add -D HL_DISABLE_LEGACY_FFI to hlc template"

This reverts commit 6b6feb56196a42c460a996e63b150c06e1daf016.

yuxiaomao committed:
Add -D HL_DISABLE_LEGACY_FFI in hlc.h instead of template

tobil4sk commented:
> Do you have other suggestions on this?

I was hoping it might be possible to just repeat the logic in both `hl.h` and `hl_ffi.h`, but seems like HL_NAME makes that difficult, because HL_NAME being unset seems to be used as a way to detect if libhl is being built. Might be good to clean up the logic a bit, that might make this easier.

Ideally `hl.h` with `HL_DISABLE_LEGACY_FFI` shouldn't set `HL_PRIM`, since that is only used for ffi.

tobil4sk committed:
Avoid redefining DEFINE_PRIM

tobil4sk committed:
Combine non-static and LIBHL_EXPORTS branch

tobil4sk committed:
Remove intermediate HL__DEFINE_PRIM_WITH_NAME

tobil4sk committed:
Define HL_PRIM in hl_ffi.h

tobil4sk committed:
Fix HL_PRIM redefinition between hl.h and hl_ffi.h

tobil4sk committed:
Use HL_EXTERN_C macro in hl.h as well

tobil4sk committed:
Allow using hl_ffi.h without HL_NAME

This means the primitive macros will not export anything.

tobil4sk committed:
Undefine C_FUNCTION_BEGIN/END for new ffi

Avoids polluting with global macros

tobil4sk committed:
Make hl_ffi.h independent of hl.h

tobil4sk committed:
Replace EXPORT/IMPORT with HL_EXPORT/HL_IMPORT

IMPORT and EXPORT are still defined by default for backwards
compatbility, unless HL_DISABLE_LEGACY_FFI is used.

tobil4sk committed:
Merge remote-tracking branch 'upstream/master' into dev-ffi

yuxiaomao committed:
Update dev-ffi branch

yuxiaomao commented:
@ncannasse please check this when you have time.

This PR add a new file `hl_ffi.h` to be used with `#define HL_DISABLE_LEGACY_FFI`, for hdll developers and avoid redefintion/undef of not prefixed defines. And to deprecated legacy ffi in the future (HL 2.0 maybe).

yuxiaomao committed:
Merge branch 'master' into dev-ffi

yuxiaomao commented:
I think it's a good first step to cleanup hl.h, so I'll merge it.

ncannasse commented:
What are we trying to solve here exactly?

yuxiaomao commented:
This PR solve: some macro in hl.h does not have `HL_` prefix, which sometimes collide with other macro, especially GUID (see linked issue). For compatibility the new definition is only available when `-D HL_DISABLE_LEGACY_FFI`.

By cleanup hl.h, I mean hl.h is a common header for hl, libhl, hdll, hlc. Some of the function are only meant to be used internally by hl/libhl, some other are only for libraries libhl/hdlls (not hlc).

ncannasse commented:
I like the idea of a single include for both HLC and embedding HL. I'm not
sure why they should be separated

Le ven. 10 avr. 2026, 15:57, Yuxiao Mao ***@***.***> a écrit :

> *yuxiaomao* left a comment (HaxeFoundation/hashlink#859)
> <https://github.com/HaxeFoundation/hashlink/pull/859#issuecomment-4224251633>
>
> This PR solve: some macro in hl.h does not have HL_ prefix, which
> sometimes collide with other macro, especially GUID (see linked issue). For
> compatibility the new definition is only available when -D
> HL_DISABLE_LEGACY_FFI.
>
> By cleanup hl.h, I mean hl.h is a common header for hl, libhl, hdll, hlc.
> Some of the function are only meant to be used internally by hl/libhl, some
> other are only for libraries libhl/hdlls (not hlc).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/HaxeFoundation/hashlink/pull/859#issuecomment-4224251633>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHZXQCTWZVQYWOAFJCNTVD4VD4TPAVCNFSM6AAAAACQMKPX26VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHM2DEMRUGI2TCNRTGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

tobil4sk commented:
`hl_ffi.h` is for defining hdll primitives, not for embedding. There isn't an overlap between hdll primitives and hlc and ideally `hlc.h` should be as minimal as possible to avoid taking up hlc symbols, which is why they are separate.

ncannasse commented:
I would then prefer a single hl.h with a define pre include HL_HLC set in
the .c/.h generated by HL/C that would only include what's necessary for
them.


Le lun. 20 avr. 2026, 13:16, tobil4sk ***@***.***> a écrit :

> *tobil4sk* left a comment (HaxeFoundation/hashlink#859)
> <https://github.com/HaxeFoundation/hashlink/pull/859?email_source=notifications&email_token=AAHZXQA2NJNT5G37ZWXPNAT4WYBIBA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTIMRYGAYTKNRUGM4KM4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJNLQOJPWG33NNVSW45C7N5YGK3S7MNWGSY3L#issuecomment-4280156438>
>
> hl_ffi.h is for defining hdll primitives, not for embedding. There isn't
> an overlap between hdll primitives and hlc and ideally hlc.h should be as
> minimal as possible to avoid taking up hlc symbols, which is why they are
> separate.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/HaxeFoundation/hashlink/pull/859?email_source=notifications&email_token=AAHZXQA2NJNT5G37ZWXPNAT4WYBIBA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTIMRYGAYTKNRUGM4KM4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJNLQOJPWG33NNVSW45C7N5YGK3S7MNWGSY3L#issuecomment-4280156438>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHZXQEEANOIFBSCXW5JXFT4WYBIBAVCNFSM6AAAAACQMKPX26VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHM2DEOBQGE2TMNBTHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
